gh-106581: Split `CALL_PY_EXACT_ARGS` into uops

gvanrossum opened the pull request:
This is only the first step for doing `CALL` in Tier 2. The next step involves tracing into the called code object. After that we'll have to do the remaining `CALL` specialization. Finally we'll have to tweak various things like `KW_NAMES`, and possibly move the `NULL` (for method calls) *above* the callable (that's 107788). But those are things for future PRs.

Note: this moves setting `frame->return_offset` directly in front of `DISPATCH_INLINED()`, to make it easier to move it into `_PUSH_FRAME`.

- [x] Rebase on gh-107788
- [x] Redo using Mark's ideas
- [x] Add back a call to `_Py_EnterRecursivePy`
- [x] Add tests


<!-- gh-issue-number: gh-106581 -->
* Issue: gh-106581
<!-- /gh-issue-number -->

gvanrossum committed:
Split `CALL_PY_EXACT_ARGS` into uops

This is only the first step for doing `CALL` in Tier 2.
The next step involves tracing into the called code object.
After that we'll have to do the remaining `CALL` specialization.
Finally we'll have to tweak various things like `KW_NAMES`,
and possibly move the `NULL` (for method calls) *above* the callable.
But those are things for future PRs.

Note: this moves setting `frame->return_offset` directly in front of
`DISPATCH_INLINED()`, to make it easier to move it into `_PUSH_FRAME`.

brandtbucher commented on Python/bytecodes.c:
```diff
@@ -2955,18 +2954,35 @@ dummy_func(
+                callable = method;
+                args--;
+                argcount++;
+            }
```
It's a shame that this logic needs to be duplicated.

Also, not sure if you saw #107788, but one of us is going to have conflicts. It's not too bad, though.

gvanrossum replied:
It's a bloody shame, but I don't know how to avoid it. We can put effort into this later if we feel it's needed.

Do you expect to be landing your PR before you head out on vacation? In that case I think you should go first. I expect that my stuff gets slightly simpler after that. I'll review it in a bit.

brandtbucher replied:
Yeah, I'd like to land mine by tomorrow if possible. Thanks.

brandtbucher commented on Python/bytecodes.c:
```diff
@@ -2955,18 +2954,35 @@ dummy_func(
+                args--;
+                argcount++;
+            }
             STAT_INC(CALL, hit);
```
For my own understanding: this is staying here, because if by some miracle to remove the `_CHECK_CALL_PY_EXACT_ARGS` uop, our stats will still be correct?

gvanrossum replied:
Hm, I suppose. It just felt to belong in the action rather than the guard.

brandtbucher commented on Python/bytecodes.c:
```diff
@@ -2955,18 +2954,35 @@ dummy_func(
             DEOPT_IF(!_PyThreadState_HasStackSpace(tstate, code->co_framesize), CALL);
+        }
+
+        op(_INIT_CALL_PY_EXACT_ARGS, (method, callable, args[oparg] -- new_frame: _PyInterpreterFrame*)) {
```
It feels weird to push `new_frame`, but I do understand that it's sort of our only way to do this move.

If we didn't need to sneak a `SET_IP` in there during tracing, then `_PUSH_FRAME` could have just been bundled into this uop, right? Hm.

gvanrossum replied:
Definitely weird (and I had to fix a few nits in the code generator to make it work) but I really want to have `_PUSH_FRAME` as a separate uop:

- There are a bunch of other instructions ending in `DISPATCH_INLINED()` and those will all be able to share `_PUSH_FRAME`.
- This uop needs to be special-cased by the (yet to be written) Tier 2 optimizer -- it's better to have fewer special cases.

brandtbucher commented on Python/executor.c:
```diff
@@ -30,6 +30,19 @@
+    do {                                                        \
+        assert(tstate->interp->eval_frame == NULL);             \
+        _PyFrame_SetStackPointer(frame, stack_pointer);         \
+        frame->prev_instr -= 1;                                 \
```
Why isn't this `next_instr - 1` anymore?

gvanrossum replied:
Because Tier 2 doesn't have a `next_instr` variable. That's also why we have to sneak in the `SAVE_IP` uop -- it sets `prev_instr`. The `_PUSH_FRAME` uop (and hence this macro) literally doesn't have enough information to know what to set it to -- the oparg provided to `SAVE_IP` is calculated by the translator (and I don't want to abuse `operand` or do other irregular things here).

brandtbucher commented on Python/optimizer.c:
```diff
@@ -611,6 +615,10 @@ translate_bytecode_to_trace(
                         ADD_TO_TRACE(expansion->uops[i].uop, oparg, operand);
+                        if (expansion->uops[i].uop == _PUSH_FRAME) {
+                            ADD_TO_TRACE(SAVE_IP, 0, 0);
+                            goto done;
```
Maybe assert that this is the last uop in the expansion (since we're bailing out of the inner loop)?

gvanrossum replied:
Good idea.
```
assert(i + 1 == nuops);
```

brandtbucher commented on Tools/cases_generator/generate_cases.py:
```diff
@@ -428,13 +429,22 @@ def add(name: str) -> None:
+                # _PUSH_FRAME is super special; it expands to SAVE_IP(next_instr) + _PUSH_FRAME
+                if part.instr.name == "_PUSH_FRAME":
+                    expansions.append(
+                        ("SAVE_IP", OPARG_SIZES["OPARG_SAVE_IP"], cache_offset)
```
Is there an elegant way to just include this in the expansion in `bytecodes.c`? Maybe not, since it's uop-specific...

gvanrossum replied:
Not really. We'd still have to special-case it here, because the "size" and "oparg" fields in the expansion are still irregular; *and* we'd have to skip it when generating the Tier 1 code for the macro, because we don't want it there.

brandtbucher commented on Tools/cases_generator/instructions.py:
```diff
@@ -120,11 +120,13 @@ def __init__(self, inst: parsing.InstDef):
-        # if self.name.startswith("CALL"):
-        #     dprint = print
+        if "PY_EXACT" in self.name:
+            dprint = print
```
Did you mean to leave this uncommented?

gvanrossum replied:
Kind of, since it should actually be silent (`CALL_PY_EXACT_ARGS` is viable now).

brandtbucher reviewed (approved):
Thanks for tackling this, it definitely doesn't look easy. It's sort of a bummer that we need to special-case this much stuff, but I also don't see a nicer way of handling these issues than what you have here.

A few comments and questions, mostly for my own understanding:

gvanrossum reviewed (commented):
I'll add that assert; then I'll review your PR, and hopefully you can then merge that, and I can handle the merge fallout.

markshannon commented on Python/bytecodes.c:
```diff
@@ -2955,18 +2954,35 @@ dummy_func(
             DEOPT_IF(!_PyThreadState_HasStackSpace(tstate, code->co_framesize), CALL);
+        }
+
+        op(_INIT_CALL_PY_EXACT_ARGS, (method, callable, args[oparg] -- new_frame: _PyInterpreterFrame*)) {
```
As this makes a frame, perhaps rename it to `_MAKE_FRAME`?

gvanrossum replied:
I figured there will be other uops making frames once we try to split CALL_PY_WITH_DEFAULTS, CALL_NO_KW_ALLOC_AND_ENTER_INIT, as well as BINARY_SUBSCR_GETITEM, LOAD_ATTR_PROPERTY, LOAD_ATTR_GETATTRIBUTE_OVERRIDDEN.

Anyway, uop names are easily changed.

markshannon commented on Python/bytecodes.c:
```diff
@@ -2955,18 +2954,35 @@ dummy_func(
-            SKIP_OVER(INLINE_CACHE_ENTRIES_CALL);
+        }
+
+        op(_PUSH_FRAME, (new_frame: _PyInterpreterFrame* -- unused)) {
```
Since `_PUSH_FRAME` is just `frame->return_offset = 0; DISPATCH_INLINED(new_frame);`, it would make sense to spell out `DISPATCH_INLINED` to clarify which operations that need to be different for tier 1 and tier 2.
Something like:
```C
op(_PUSH_FRAME, (new_frame: _PyInterpreterFrame* -- unused)) {
    SAVE_FRAME_STATE();   // Equivalent to frame->prev_instr = next_instr - 1; _PyFrame_SetStackPointer(frame, stack_pointer);
    frame->return_offset = 0;
    new_frame->previous = frame;
    frame = cframe.current_frame = new_frame;
    CALL_STAT_INC(inlined_py_calls);
    if (_Py_EnterRecursivePy(tstate)) {
        goto exit_unwind;
    }
    START_FRAME(); // Equivalent to next_instr = frame->prev_instr  + 1; stack_pointer = 
            stack_pointer = _PyFrame_GetStackPointer(frame);
}
```

markshannon replied:
For example: https://github.com/python/cpython/commit/2b3e6f2f0a5c3fcb7449b2e3bfe21d44c7554870

In which case the code generators needs to know to push the temporary stack values to the real stack before `SAVE_FRAME_STATE()`

gvanrossum replied:
I have to study this more. A problem is that the Tier 1 and Tier 2 versions of `_PUSH_FRAME` are so different. I am working on a mechanism to be able to say
```
#if TIER_ONE
<code for Tier 1>
#else
<code for Tier 2>
#endif
```
I'm not sure yet what you mean with your last remark about pushing temp stack values.

gvanrossum replied:
Comparing carefully the two versions of `DISPATCH_INLINED` (adding `frame->return_offset = 0` which precedes it in both cases):

In Tier 1:
```
        frame->return_offset = 0;
        assert(tstate->interp->eval_frame == NULL);
        _PyFrame_SetStackPointer(frame, stack_pointer);
        frame->prev_instr = next_instr - 1;
        (NEW_FRAME)->previous = frame;
        frame = cframe.current_frame = (NEW_FRAME);
        CALL_STAT_INC(inlined_py_calls);
        goto start_frame;
```

In Tier 2:
```
        frame->return_offset = 0;
        assert(tstate->interp->eval_frame == NULL);
        _PyFrame_SetStackPointer(frame, stack_pointer);
        frame->prev_instr -= 1;
        (NEW_FRAME)->previous = frame;
        frame = tstate->cframe->current_frame = (NEW_FRAME);
        CALL_STAT_INC(inlined_py_calls);
        stack_pointer = _PyFrame_GetStackPointer(frame);
        ip_offset = (_Py_CODEUNIT *)_PyFrame_GetCode(frame)->co_code_adaptive;
```

Diff:
```diff
@@ -1,8 +1,9 @@
         frame->return_offset = 0;
         assert(tstate->interp->eval_frame == NULL);
         _PyFrame_SetStackPointer(frame, stack_pointer);
-        frame->prev_instr = next_instr - 1;
+        frame->prev_instr -= 1;
         (NEW_FRAME)->previous = frame;
-        frame = cframe.current_frame = (NEW_FRAME);
+        frame = tstate->cframe->current_frame = (NEW_FRAME);
         CALL_STAT_INC(inlined_py_calls);
-        goto start_frame;
+        stack_pointer = _PyFrame_GetStackPointer(frame);
+        ip_offset = (_Py_CODEUNIT *)_PyFrame_GetCode(frame)->co_code_adaptive;
```

Note that the Tier 2 version must be preceded by a `SAVE_IP`, which does the equivalent of `frame->prev_instr = next_instr`. If we had a Tier 1 version of `SAVE_IP` we could include it in the macro definition:
```
        macro(CALL_PY_EXACT_ARGS) =
            unused/1 + // Skip over the counter
            _CHECK_PEP_523 +
            _CHECK_FUNCTION_EXACT_ARGS +
            _CHECK_STACK_SPACE +
            _INIT_CALL_PY_EXACT_ARGS +
            SAVE_IP +      // <-------------- added
            _PUSH_FRAME;
```
which would reduce the special-casing in the code generator a bit (it would still need to do something special for `SAVE_IP` to ensure that its `oparg` has the right value, different from the `oparg` of the macro (which is the argument count). This would take care of the first diff chunk (what to assign to `frame->prev_inst`), but it would still be pretty fragile. (Like my current version, it would entice the optimizer to incorrectly try to remove the `SAVE_IP` uop.)

The second diff chunk relates to how we set `cframe.current_frame` -- in Tier 2 we must access this through the `tstate`.

The third and final diff chunk relates to really start using the new frame. In Tier 1, this must actually do the following:
- Check recursion
- Load `stack_pointer`
- Load `next_instr`
- Dispatch to the next opcode.

This is done by the code at `start_frame`.

In Tier 2 there is no `start_frame` label (the only uop that can go to a label is `EXIT_TRACE`, and of course `DEOPT_IF` and `ERROR_IF` also jump). So we load `stack_frame` here. There is no direct equivalent to `next_instr`, but we have to set `ip_offset`, which `SAVE_IP` adds to its `oparg` to get the `prev_instr` value. (This variable is a cache for `frame->code->co_code_adaptive`, to save some memory loads, so whenever `frame` changes we must update it.)

(More later.)

gvanrossum replied:
There's another thing though, and I think that is what Mark meant. In Tier 1 the code generation for macros is special-cased for `_PUSH_FRAME` so that both the stack adjustment *and* the `next_instr` adjustment are emitted *before* the `_PUSH_FRAME` opcode. This is done so that the flushing of these variables to the frame in the `DISPATCH_INLINED` macro flush the correct values.

But this is really ugly and unprincipled, and the logic is much hairier than the other special cases for `_PUSH_FRAME`. One of Mark's ideas here is to make this special case look for uops using the `SAVE_FRAME_STATE` macro rather than for the specific uop `_PUSH_FRAME`. But detecting when to trigger the special case is only part of the problem -- IMO the worse problem is that the special case itself is so ugly:
```
        dispatch_inlined_special_case = False
        if mgr is managers[-1] and mgr.instr.always_exits.startswith("DISPATCH_INLINED") and mgr.instr.name == "_PUSH_FRAME":
            dispatch_inlined_special_case = True
            temp = mgr.final_offset.clone()
            temp.deeper(StackEffect(UNUSED))  # Hack
            out.stack_adjust(temp.deep, temp.high)
            # Use clone() since adjust_inverse() mutates final_offset
            mgr.adjust_inverse(mgr.final_offset.clone())
            if cache_adjust:
                out.emit(f"next_instr += {cache_adjust};")
```
The last 4 lines here, starting with `# Use clone()`, occur further down too, for the normal case (*after* the final uop). I don't even recall why the `temp.deeper()` call is needed!

I'll mull this over some more.

gvanrossum replied:
I think I have addressed this. @markshannon Please have another look. Assuming the tests look okay I'll un-draft this.

brandtbucher commented on Python/bytecodes.c:
```diff
@@ -2955,18 +2954,35 @@ dummy_func(
             PyCodeObject *code = (PyCodeObject *)func->func_code;
             DEOPT_IF(code->co_argcount != argcount, CALL);
             DEOPT_IF(!_PyThreadState_HasStackSpace(tstate, code->co_framesize), CALL);
```
Looking at this again, I think that this particular "guard" should actually be part of the action half.

The purpose of splitting things up is to remove or reorder guards, but this particular check for stack space must occur immediately before the frame push, and can never be skipped (unlike the other checks in this uop, which may be statically proven pretty easily). So it would likely prevent optimizations around calls in the future (or worse, we might forget accidentally remove it during optimization).

I know we can't do much with it now, but it seems worth thinking about these things carefully while we're here in the early stages.

gvanrossum replied:
Made into a separate guard (a compromise we discussed this morning).

gvanrossum committed:
Fix merge so it works again (I think)

gvanrossum committed:
Split into finer-grained uops

gvanrossum committed:
Fix type error in stacking.py

gvanrossum committed:
Add test

gvanrossum committed:
Add comment explaining _PUSH_FRAME's unused output effect

gvanrossum committed:
Make PUSH_FRAME special case a little less myterious

gvanrossum committed:
Rename Instruction.write to write_case_body

gvanrossum committed:
Move next_instr update to a more logical place

gvanrossum committed:
Don't recompute macro cache offset

gvanrossum committed:
Fold and refactor long line in stacking.py

gvanrossum committed:
Fold long lines in generate_cases.py

gvanrossum committed:
Don't emit static assert to executor cases

gvanrossum committed:
Factor away write_case_body (formerly Instruction.write)

gvanrossum committed:
Make less of a special case of _PUSH_FRAME

Instead, the special case is an opcode using SAVE_FRAME_STATE().
Introducing #if TIER_ONE and #if TIER_TWO so we can implement
_PUSH_FRAME differently for both tiers.

gvanrossum committed:
Stop special-casing _PUSH_FRAME altogether

Instead, we special-case SAVE_IP:
- Its Tier 2 expansion sets oparg to the instruction offset
- In Tier 1 it is a no-op (and skipped if present in a macro)

gvanrossum committed:
Fold long lines

gvanrossum committed:
Call _Py_EnterRecursivePy in _FRAME_PUSH

gvanrossum committed:
Merge remote-tracking branch 'upstream/main' into call-uops

markshannon commented on Python/executor.c:
```diff
@@ -30,6 +30,14 @@
 #define ENABLE_SPECIALIZATION 0
 
+#undef SAVE_FRAME_STATE
+#define SAVE_FRAME_STATE() \
```
Rather than a macro, I think the code generator needs to understand this.

Given that `SAVE_FRAME_STATE` is basically `SAVE_CURRENT_IP` followed by `_PyFrame_SetStackPointer(frame, stack_pointer);` we could convert it to two micro-ops: `SAVE_CURRENT_IP` and `SAVE_SP`.

In general, we want to avoid macros in the generated C code.
The generated code can be explicit and verbose.

gvanrossum replied:
Well, there are already many macros (and static inline functions) in the generated code. The generator *recognizes* the presence of `SAVE_FRAME_STATE()`, but it doesn't *expand it* -- the C preprocessor can do that for us more easily. Currently we only do the expansion in the generator for things whose expansion requires information that only the generator has (like the stack adjustment for `ERROR_IF`).

gvanrossum replied:
You are proposing that the macro expansion for `CALL_PY_EXACT_ARGS` become
```
        macro(CALL_PY_EXACT_ARGS) =
            unused/1 + // Skip over the counter
            _CHECK_PEP_523 +
            _CHECK_FUNCTION_EXACT_ARGS +
            _CHECK_STACK_SPACE +
            _INIT_CALL_PY_EXACT_ARGS +
            SAVE_IP +  // Tier 2 only; special-cased oparg
            SAVE_CURRENT_IP +  // <------------------- Addition
            _PUSH_FRAME;
```
where `SAVE_CURRENT_IP` is something like this:
```
op(SAVE_CURRENT_IP, (--)) {
    #if TIER_ONE
    frame->prev_instr = next_instr - 1;
    #endif
    #if TIER_TWO
    frame->prev_instr--;
    #endif
}
```
Or we could special-case its expansion in the generator, potayto-potato. But it *has* to differ between tiers because in Tier 1 it must store next_instr whereas in Tier 2 it must rely on the preceding SAVE_IP to set `frame->prev_instr`. (Ideally at some point in the future we won't need the `prev_instr--` yet, but that's a tricky change.)

The `_PyFrame_SetStackPointer(frame, stack_pointer);` call should be moved back into `_PUSH_FRAME` (at the point where I currently call `SAVE_FRAME_STATE`).

If I can get this to work I'll apply it and merge the PR.

gvanrossum replied:
I *did* get this working (see https://github.com/python/cpython/pull/107760/commits/05af84805a04747c1de7a5b843041b8d55c31bee), and will test and benchmark it before merging it.

Note that there are still some `#if TIER_ONE` and `#if TIER_TWO` sections, but they are unavoidable.

markshannon commented on Python/ceval_macros.h:
```diff
@@ -103,11 +103,16 @@
         DISPATCH_GOTO(); \
     }
 
+#define SAVE_FRAME_STATE() \
```
See my comment below about splitting this into `SAVE_CURRENT_IP; SAVE_SP`

markshannon reviewed (commented):
I'm uneasy about the introduction of the `TIER_ONE` and `TIER_TWO` macros.
It is a principle of the overall design that there is a single source of truth for the semantics of bytecodes.

It might appear that I'm being dogmatic, but the need for something like those macros often indicates an underlying problem that should be fixed independently.

In this case the problem is the `cframe`. Loading and saving the IP needs to handled specially anyway and saving and loading the SP should be the same for both interpreters (but will need to be handled specially by the copy-and-patch compiler, so should be its own micro-op).
It is pushes the frame that differs. Removing `cframe` will fix that.

The `cframe` only exists as a performance hack to minimize the impact of tracing prior to PEP 669.

gvanrossum committed:
Introduce SAVE_CURRENT_IP uop per Mark's proposal
